src/profile-logic/import/art-trace.js

```diff
@@ -0,0 +1,932 @@
+  }
+
+  exitMethod(_methodId) {
+    if (this._currentStack === null) {
+      // This has been observed to happen in tracing-based traces (rather than sampling-based traces). Not sure why.
+      // console.warn('exiting method when stack is empty');
+      return;
+    }
+    this._currentStack = this._stackTable.data[this._currentStack][1];
+  }
+
+  // Called before enter/exitMethod are called for this time
+  advanceTimeTo(timestampInMSSinceStartTime) {
+    if (this._nextSampleTimestamp === 0) {
+      this._nextSampleTimestamp = timestampInMSSinceStartTime;
+      if (this._name !== 'main') {
+        this._registerTime = timestampInMSSinceStartTime;
+      }
+      return;
+    }
+
+    // Write out the previous sample, and synthesize samples in between the
+    // two timestamps. Unfortunately the ART trace format does not record
+    // timestamps for samples in which the stack does not change. We fill up
+    // such gaps with evenly-spaced synthesized samples at the interval that
+    // was divined earlier.
+    while (this._nextSampleTimestamp < timestampInMSSinceStartTime) {
+      this._samples.data.push([this._currentStack, this._nextSampleTimestamp]);
+      if (this._honorOriginalSamplingTimestamps) {
+        // Only use this loop to fill up any gaps that are at least 2 * interval wide.
+        // When less than 2 * interval remains, snap to the original next timestamp.
+        if (
+          timestampInMSSinceStartTime - this._nextSampleTimestamp <
+          2 * this._intervalInMsec
+        ) {
+          this._nextSampleTimestamp = timestampInMSSinceStartTime;
+          break;
+        }
+      }
+      this._nextSampleTimestamp += this._intervalInMsec;
+    }
+  }
+
+  finish() {
+    if (this._nextSampleTimestamp !== null) {
+      this._samples.data.push([this._currentStack, this._nextSampleTimestamp]);
+    }
+    return {
+      tid: this._tid,
+      pid: this._pid,
+      name: this._name,
+      registerTime: this._registerTime,
+      unregisterTime: null,
+      markers: this._markers,
+      samples: this._samples,
+      frameTable: this._frameTable,
+      stackTable: this._stackTable,
+      stringTable: this._stringTable,
+    };
+  }
+}
+
+// Don't try and type this more specifically. It will be run through the Gecko upgrader
+// process.
+type GeckoProfileVersion11 = Object;
+
+export function isArtTraceFormat(traceBuffer: ArrayBuffer) {
+  return detectArtTraceFormat(traceBuffer) !== 'unrecognized';
+}
+
+// Convert an ART trace to the Gecko profile format.
+export function convertArtTraceProfile(
+  traceBuffer: ArrayBuffer
+): GeckoProfileVersion11 {
+  const trace = parseArtTrace(traceBuffer);
+  const originalIntervalInUsec = procureSamplingInterval(trace);
+  let honorOriginalSamplingTimestamps = false;
+  let intervalInMsec;
+  if (!originalIntervalInUsec || originalIntervalInUsec < 100) {
+    // This is probably a tracing-based profile. Do fake-sampling.
```

Comment:
Optional: The processed profile format can do real tracing: https://github.com/firefox-devtools/profiler/blob/2d503ed75fc6ccd4d2261a7ddd65e14eb50cff37/src/types/profile.js#L111

I'd be fine copying the the types out for a processed profile version and fixing it at a version for an importer.

Reply:
Ah. I'm going to leave this one for somebody to fix in a follow-up.